File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.Calculator

type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    args.Length = 3

let parseOperation (arg : string) =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | _ -> InvalidOperationException "Операция не распознана. Вводите только +, -, *, /" |> raise
    
let parseNum (arg: string) =
    let mutable result = 0.0
    if Double.TryParse(arg, &result) then result
    else ArgumentException "Пожалуйста используйте только цифры" |> raise
        
    
let parseCalcArguments(args : string[]) =
    if isArgLengthSupported args then {
        arg1 = parseNum args[0]
        operation =  parseOperation args[1]
        arg2 = parseNum args[2]
    }
    else ArgumentException "Неверное количество аргументов. Введите число, одну из операций +, -, *, /, число" |> raise

File: Homework8/Hw8/Calculator/Parser.cs
using System.Globalization;

namespace Hw8.Calculator;

public static class Parser
{
    private static void ParseOperation(string operation, out Operation op)
    {
        op = operation.ToLower() switch
        {
            "plus" => Operation.Plus,
            "minus" => Operation.Minus,
            "multiply" => Operation.Multiply,
            "divide" => Operation.Divide,
            _ => throw new InvalidOperationException(Messages.InvalidOperationMessage)
        };
    }

    public static void ParseArgs(string arg1, string operation, string arg2,
        out double val1, out Operation op, out double val2)
    {
        if (!Double.TryParse(arg1, NumberStyles.Float, CultureInfo.InvariantCulture, out val1) 
            || !Double.TryParse(arg2, NumberStyles.Float, CultureInfo.InvariantCulture, out val2))
            throw new ArgumentException(Messages.InvalidNumberMessage);
        ParseOperation(operation, out op);
    }
}